dds_web/api/project.py

```diff
@@ -149,6 +149,75 @@ def create_project(self, data, **kwargs):
 ####################################################################################################
 # ENDPOINTS ############################################################################ ENDPOINTS #
 ####################################################################################################
+class ProjectStatus(flask_restful.Resource):
+    """Get and update Project status"""
+
+    @auth.login_required
+    def get(self):
+        """Get current project status and optionally entire status history"""
+        project = project_schemas.ProjectRequiredSchema().load(flask.request.args)
+        extra_args = flask.request.json
+        return_info = {"current_status": project.current_status}
+
+        if extra_args and extra_args.get("history") == True:
+            history = []
+            for pstatus in project.project_statuses:
+                history.append(tuple((pstatus.status, pstatus.date_created)))
+            history.sort(key=lambda x: x[1])
+            return_info.update({"history": history})
+
+        return flask.jsonify(return_info)
+
+    @auth.login_required(role=["Super Admin", "Unit Admin", "Unit Personnel"])
+    def post(self):
+        """Update Project Status"""
+        project = project_schemas.ProjectRequiredSchema().load(flask.request.args)
+        extra_args = flask.request.json
+        new_status = extra_args.get("new_status")
+        if new_status not in [
+            "In Progress",
+            "Deleted",
+            "Available",
+            "Expired",
+            "Archived",
+        ]:
+            raise DDSArgumentError("Invalid status")
+
+        if not self.is_transition_possible(project.current_status, new_status):
+            raise DDSArgumentError("Invalid status transition")
+
+        try:
+            add_status = models.ProjectStatuses(
+                **{
+                    "project_id": project.id,
+                    "status": new_status,
+                    "date_created": dds_web.utils.current_time(),
+                }
+            )
+
+            project.project_statuses.append(add_status)
+            db.session.commit()
+        except (sqlalchemy.exc.SQLAlchemyError, TypeError) as err:
+            flask.current_app.logger.exception(err)
+            db.session.rollback()
+            raise DatabaseError(message="Server Error: Status was not updated")
+
+        return flask.jsonify({"message": f"{project.public_id} updated to status {new_status}"})
+
+    def is_transition_possible(self, current_status, new_status):
+        """Check if the transition is valid"""
+        possible_transitions = [
+            ("In Progress", ["Available", "Deleted"]),
+            ("Available", ["In Progress", "Expired", "Archived"]),
+            ("Expired", ["Available", "Archived"]),
+        ]
+        result = False
+        if current_status != new_status:
```

Comment:
Maybe we could add something like this
```
if current_status == new_status: 
    raise DDSArgumentError(f"The project status is already {new_status}")
```
and then underneath
```
for transition in possible_transitions:
    .... etc
```
Thoughts?

Reply:
Maybe we don't need the check at all as it is not defined in `possible_transitions`? It will then then flag it as invalid and raise `DDSArgumentError("Invalid status transition")` in the calling function.

Reply:
True, do that instead!